sycl/source/detail/context_impl.cpp

```diff
@@ -87,11 +87,25 @@ context_impl::~context_impl() {
     PI_CALL(piContextRelease)(m_Context);
   }
   // Release all programs and kernels created with this context
-  for (auto ProgIt : m_CachedPrograms) {
-    RT::PiProgram ToBeDeleted = ProgIt.second;
-    for (auto KernIt : m_CachedKernels[ToBeDeleted]) {
-      PI_CALL(piKernelRelease)(KernIt.second);
+  for (auto &ProgIt : m_CachedPrograms) {
+    WithBuildState<PiProgramT> &ProgWithState = ProgIt.second;
+    PiProgramT *ToBeDeleted = ProgWithState.Ptr.load();
+
+    if (!ToBeDeleted) {
+      continue;
     }
+
+    auto KernIt = m_KernelsPerProgramCache.find(ToBeDeleted);
+
+    if (KernIt != m_KernelsPerProgramCache.end())
```

Comment:
Maybe for each loop?

Reply:
There is already a foreach loop.
Can;t use map's `[]` operator due to it has special requirements for construction of map's value objects which `atomic` won't ever meet.